### ISSUE TITLE
Setting es.read.field.include breaks mapping discovery for fields nested deeper than two levels

### ISSUE THREAD

**Issue by jbaiera** [elastic/elasticsearch-hadoop#1139]
When read field include is used with a field nested more than two levels deep, the absolute path of the parent name is not considered for filtering (only the simple name of the parent). In the case of:

```json
"grandparent": {
  "parent": {
    "child": "blah"
  }
}
```
and setting `es.read.field.include` to `grandparent.parent.child` leads to an error:
```
org.elasticsearch.hadoop.EsHadoopIllegalStateException: Field 'parent.field1' not found; typically this occurs with arrays which are not mapped as single value
	at org.elasticsearch.spark.sql.RowValueReader$class.rowColumns(RowValueReader.scala:51)
	at org.elasticsearch.spark.sql.ScalaRowValueReader.rowColumns(ScalaEsRowValueReader.scala:32)
	at org.elasticsearch.spark.sql.ScalaRowValueReader.createMap(ScalaEsRowValueReader.scala:69)
```
What is happening here is that when we discover the mapping from Elasticsearch, we filter all of the fields on that mapping to see if they match the include/exclude settings. This is done recursively, and constructs the field name to be filtered by using the name of the parent field. Unfortunately, it seems that the parent's absolute name is not used when filtering, only its simple name. This means that the filter attempts to check if `parent.child` should be included instead of `grandparent.parent.child`, thus failing the inclusion check and the data is removed.

This leads to the field information being missing from the serialization code. The serialization code correctly assumes to include the field based on the correct inclusion logic, but the field information cannot be found because it was incorrectly removed from the mapping when it was discovered.

**Comment by markpavey** [comment]
@jbaiera Would it be possible to have this fix back-ported to the 5.6 branch? I am using Elasticsearch 5.6 and have just lost quite a bit of time due to this bug, despite upgrading to the latest 5.6.x version.

I have found a workaround, which is to specify "parent.child" as well as "grandparent.parent.child" in es.read.field.include but I would prefer not to have to do this.

Many thanks!

**Comment by jbaiera** [comment]
@markpavey this is backported now - I had the incorrect assumption when it was first fixed that the overhauled mapping code was on the 6.x line only, but now remember that it was actually backported to 5.6. Sometimes looking at the history of a file is not always the whole picture. The fix should be in the next 5.6 release. Sorry for the inconvenience!.

**Comment by markpavey** [comment]
@jbaiera no worries, thanks for making the change!
